telethon_examples/payment.py

```diff
@@ -0,0 +1,143 @@
+@bot.on(events.Raw(types.UpdateBotPrecheckoutQuery))
+async def payment_pre_checkout_handler(event: types.UpdateBotPrecheckoutQuery):
+    if event.payload.decode("UTF-8") == 'product A':
+        #  so we have to confirm payment
+        await bot(
+            functions.messages.SetBotPrecheckoutResultsRequest(
+                query_id=event.query_id,
+                success=True,
+                error=None
+            )
+        )
+    elif event.payload.decode("UTF-8") == 'product B':
+        # same for another
+        await bot(
+            functions.messages.SetBotPrecheckoutResultsRequest(
+                query_id=event.query_id,
+                success=True,
+                error=None
+            )
+        )
+    else:
+        # for example, something went wrong (whatever reason). We can tell customer about that:
+        await bot(
+            functions.messages.SetBotPrecheckoutResultsRequest(
+                query_id=event.query_id,
+                success=False,
+                error="Something went wrong"
+            )
+        )
+
+    raise events.StopPropagation
+
+
+# That event is handled at the end, when customer payed.
+@bot.on(events.Raw(types.UpdateNewMessage))
+async def payment_received_handler(event):
+    if isinstance(event.message.action, types.MessageActionPaymentSentMe):
+        payment: types.MessageActionPaymentSentMe = event.message.action
+        # do something after payment was recieved
+        if payment.payload.decode("UTF-8") == 'product A':
+            await bot.send_message(event.message.from_id, "Thank you for buying product A!")
+        elif payment.payload.decode("UTF-8") == "product B":
+            await bot.send_message(event.message.from_id, "Thank you for buying product B!")
+        raise events.StopPropagation
+
+
+# let's put it in one function for more easier way
+def generate_invoice(price_label: str, price_amount: int, currency: str, title: str,
+                     description: str, payload: str, start_param: str) -> types.InputMediaInvoice:
+    price = types.LabeledPrice(label=price_label, amount=price_amount)  # label - just a text, amount=10000 means 100.00
+    invoice = types.Invoice(
+        currency=currency,  # currency like USD
+        prices=[price],  # there could be a couple of prices.
+        test=True,  # if you're working with test token
+
+        #  next params are saying for themselves
+        name_requested=False,
+        phone_requested=False,
+        email_requested=False,
+        shipping_address_requested=False,
+        flexible=False,
+        phone_to_provider=False,
+        email_to_provider=False
+
+    )
+    return types.InputMediaInvoice(
+        title=title,
+        description=description,
+        invoice=invoice,
+        payload=payload.encode("UTF-8"),  # payload, which will be sent to next 2 handlers
+        provider=provider_token,
+        provider_data=types.DataJSON("{}"),  # honestly, no idea.

```

Comment:
I would really like it if we didn't say this. Maybe Telegram's bot API page or PTB docs have more info? It would make sense if it was provider-provided data they need, but better check.

Reply:
Actually found that one
JSON-encoded data about the
                invoice, which will be shared with the payment provider. A detailed description of
                required fields should be provided by the payment provider. When an object is
                passed, it will be encoded as JSON.